File: fluid/src/api/evolve.ts
/**
 * The consolidation queue, as `GET /evolve` answers it: what the registered
 * domains need next, ranked, with the evidence behind each finding and the
 * instruction that says how to work it.
 *
 * The sweep is a read and only a read. The `evolve` MCP tool records that a
 * sweep happened, because an agent calling it is about to do the work; this
 * endpoint runs the detection half instead, so a page left open never tells
 * the instance its backlog was attended to. The two acknowledgment verbs below
 * it are the exception that proves it: they write, and they write only when
 * somebody presses the button that calls them.
 *
 * Two shapes of the payload are worth naming here rather than at the screen.
 * A finding's `class` decides how it is drawn - `mechanical` completes intent
 * the knowledge already records, `judgment` changes what it claims - so a class
 * this client has never heard of is read as `judgment`: failing toward asking a
 * person is the only safe direction. And a queue row carries no family of its
 * own. The engine counts families over the whole filtered result and numbers
 * every rule by family in its catalog (`V0xx` temporal, `V1xx` structure,
 * `V2xx` redundancy), so the section a row belongs under is read off its rule
 * id, and a rule id from a newer catalog belongs to no section rather than to
 * the wrong one.
 */

import { api, encodeSegment } from "./client";
import { asArray, asNumber, asObject, asString } from "./json";
import type { AckBody } from "./model";

/** The detector families, in the catalog's own order. */
export const EVOLVE_FAMILIES = ["temporal", "structure", "redundancy"] as const;

/** One detector family. */
export type EvolveFamily = (typeof EVOLVE_FAMILIES)[number];

/** How a family is titled where a reader sees it. */
export const EVOLVE_FAMILY_TITLES: Record<EvolveFamily, string> = {
  temporal: "Temporal",
  structure: "Structure",
  redundancy: "Redundancy",
};

/** What a family section says it is about, in one line. */
export const EVOLVE_FAMILY_BLURBS: Record<EvolveFamily, string> = {
  temporal: "Validity windows, staleness and the supersede lifecycle.",
  structure: "References, reciprocity, orphans, stubs and size.",
  redundancy: "Duplicate content, colliding titles and tag drift.",
};

/**
 * What kind of work a finding is.
 *
 * `mechanical` completes intent the knowledge already records, so an agent may
 * just do it. `judgment` changes what the knowledge claims, so it is a question
 * for a person rather than a change to apply.
 */
export type EvolveClass = "mechanical" | "judgment";

/** What an unrecognized class reads as: the one that asks before acting. */
export const DEFAULT_EVOLVE_CLASS: EvolveClass = "judgment";

/** How many findings one sweep asks for. The engine clamps anything above it. */
export const EVOLVE_LIMIT = 100;

/**
 * How long a sweep stays fresh, in ms.
 *
 * Detection is the heaviest read this API has - it reads every engram of every
 * registered domain - and what it answers is a backlog, which moves at the
 * speed of somebody editing the knowledge rather than at the speed of a screen.
 * A minute is what the tree already uses for the same kind of value
 * (`TREE_STALE_TIME` in `api/domain.ts`), and it is what makes the core loop of
 * the maintenance screen cheap: following a finding to its engram and coming
 * back is a remount, and a remount inside the window reads the cache instead of
 * sweeping again.
 */
export const EVOLVE_STALE_MS = 60_000;

/** One finding, ranked across the whole result. */
export interface EvolveFinding {
  /** Its rank across the whole result, not within the page. */
  n: number;
  /** 0 to 100, after the salience and hub boosts. */
  priority: number;
  /** The rule that fired, for example `V006`. */
  rule: string;
  class: EvolveClass;
  domain: string;
  /**
   * The engram it fired on, or `""` for a finding that names no engram at all
   * - an orphaned attachment, a domain's drifted tag vocabulary. Empty is a
   * shape rather than a missing field: those findings are about the domain,
   * and an acknowledgment, which lives on an engram, has nowhere to hang.
   */
  permalink: string;
  /**
   * Its subject, in the words a reader knows it by: the engram's title, the
   * attachment's path for a finding about a file, and the domain's own name
   * for one that carries neither.
   */
  title: string;
  /** The line the rule fired on, or null when the finding is about the whole. */
  line: number | null;
  /** What was found, in a few words. */
  finding: string;
  /** What the detector read to find it. */
  evidence: string;
  /** What would settle it, in a few words. */
  fix: string;
  /**
   * An acknowledgment matched, and this row is here only because the sweep
   * was asked for the silenced ones.
   */
  acknowledged: boolean;
  /**
   * An acknowledgment for this rule exists on the engram but was given for
   * evidence that has since changed, so it no longer silences anything. The
   * finding is drawn as usual, saying so.
   */
  ackStale: boolean;
  /** The note the matching or stale acknowledgment carries, when it has one. */
  ackNote: string | null;
}

/** What acknowledgments kept out of the queue. */
export interface EvolveAcknowledged {
  /** Every silenced finding, over the whole result. */
  total: number;
  /** The same count per family, keyed as the engine names them. */
  byFamily: Record<string, number>;
}

/** One sweep, as this app reads it. */
export interface EvolveQueue {
  /** How many engrams the sweep read. */
  engramsScanned: number;
  /** How many findings the whole filtered result holds, page or no page. */
  total: number;
  /** Counts over the whole result rather than this page. */
  families: { family: string; findings: number }[];
  /** This page of the ranked queue. */
  queue: EvolveFinding[];
  /** The prescribed action, once per rule on this page. */
  actions: { rule: string; instruction: string }[];
  /** Any per-domain cap that fired, so a short queue is never mistaken for a
   * finished one. */
  truncations: string[];
  /**
   * What acknowledgments silenced, counted whether or not this sweep asked to
   * see it. A queue never shrinks quietly: what somebody ruled intentional is
   * still said out loud, as a number with a way to look at it.
   */
  acknowledged: EvolveAcknowledged;
}

/**
 * The family a rule belongs to, read off its id.
 *
 * Null for anything this client does not recognize, which includes a rule from
 * a catalog newer than it: a finding filed under the wrong heading would read
 * as a claim about what kind of work it is.
 */
export function evolveFamily(rule: string): EvolveFamily | null {
  const match = /^V(\d)\d\d$/.exec(rule);
  const index = Number(match?.[1] ?? NaN);
  return EVOLVE_FAMILIES[index] ?? null;
}

/** Read one class, falling back to the one that asks before acting. */
function readClass(value: unknown): EvolveClass {
  return value === "mechanical" ? "mechanical" : DEFAULT_EVOLVE_CLASS;
}

/**
 * Read one finding, or null when it names neither a rule nor a domain.
 *
 * Only those two are required. A permalink is NOT: several rules are about a
 * domain rather than about any one engram - an orphaned attachment, a drifted
 * tag vocabulary - and they arrive with an empty one by design. Requiring it
 * dropped exactly those rows while the engine went on counting them in
 * `total`, so the queue said one number and drew a smaller one.
 */
function readFinding(value: unknown): EvolveFinding | null {
  const record = asObject(value);
  const rule = asString(record?.rule);
  const domain = asString(record?.domain);
  if (rule === null || domain === null) {
    return null;
  }
  const permalink = asString(record?.permalink) ?? "";
  return {
    n: asNumber(record?.n) ?? 0,
    priority: asNumber(record?.priority) ?? 0,
    rule,
    class: readClass(record?.class),
    domain,
    permalink,
    // The attachment rules put the path in the title, so an anchorless finding
    // usually names its own subject. The one that does not - tag drift - is
    // about the domain, which is then the truest subject there is.
    title: asString(record?.title) ?? (permalink === "" ? domain : permalink),
    line: asNumber(record?.line),
    finding: asString(record?.finding) ?? "",
    evidence: asString(record?.evidence) ?? "",
    fix: asString(record?.fix) ?? "",
    // Both flags are omitted rather than sent false, so their absence is the
    // ordinary case and only a literal `true` means anything.
    acknowledged: record?.acknowledged === true,
    ackStale: record?.ack_stale === true,
    ackNote: asString(record?.ack_note),
  };
}

/** Read what acknowledgments silenced, whole and per family. */
function readAcknowledged(value: unknown): EvolveAcknowledged {
  const record = asObject(value);
  const byFamily: Record<string, number> = {};
  for (const [family, count] of Object.entries(
    asObject(record?.by_family) ?? {},
  )) {
    const findings = asNumber(count);
    if (findings !== null) {
      byFamily[family] = findings;
    }
  }
  return { total: asNumber(record?.total) ?? 0, byFamily };
}

/** Read one family count, or null when either half is missing. */
function readFamilyCount(
  value: unknown,
): { family: string; findings: number } | null {
  const record = asObject(value);
  const family = asString(record?.family);
  const findings = asNumber(record?.findings);
  return family === null || findings === null ? null : { family, findings };
}

/** Read one per-rule instruction, or null when either half is missing. */
function readAction(
  value: unknown,
): { rule: string; instruction: string } | null {
  const record = asObject(value);
  const rule = asString(record?.rule);
  const instruction = asString(record?.instruction);
  return rule === null || instruction === null ? null : { rule, instruction };
}

/** Read a sweep payload. */
export function readEvolveQueue(payload: unknown): EvolveQueue {
  const record = asObject(payload);
  return {
    engramsScanned: asNumber(record?.engrams_scanned) ?? 0,
    total: asNumber(record?.total) ?? 0,
    families: asArray(record?.families)
      .map(readFamilyCount)
      .filter(
        (count): count is { family: string; findings: number } =>
          count !== null,
      ),
    queue: asArray(record?.queue)
      .map(readFinding)
      .filter((finding): finding is EvolveFinding => finding !== null),
    actions: asArray(record?.actions)
      .map(readAction)
      .filter(
        (action): action is { rule: string; instruction: string } =>
          action !== null,
      ),
    truncations: asArray(record?.truncations).filter(
      (entry): entry is string => typeof entry === "string",
    ),
    acknowledged: readAcknowledged(record?.acknowledged),
  };
}

/**
 * Every cached sweep, whatever it was asked for: the prefix a write
 * invalidates, so acknowledging something refreshes the sweep that hides the
 * silenced rows AND the one that shows them.
 */
export const EVOLVE_KEY_ROOT = ["evolve"] as const;

/**
 * The cache key of one sweep, which is every parameter it carries.
 *
 * Whether the silenced findings were asked for is one of them: it is a
 * different question with a different answer, and reading one back for the
 * other would draw a queue that does not match the toggle above it.
 */
export function evolveKey(
  domains: string[] = [],
  includeAcknowledged = false,
): readonly unknown[] {
  return [...EVOLVE_KEY_ROOT, domains, includeAcknowledged];
}

/**
 * Fetch one sweep.
 *
 * A full page by default rather than the engine's own ten: this screen draws
 * the queue in one go, grouped by family, and paging a report somebody opened
 * to see the shape of the backlog would hide exactly the thing they came for.
 * `total` says how much a fired cap left out.
 */
export async function fetchEvolveQueue(
  opts: {
    domains?: string[];
    limit?: number;
    /** Ask for the silenced findings too, each marked acknowledged. */
    includeAcknowledged?: boolean;
  } = {},
): Promise<EvolveQueue> {
  const query = new URLSearchParams();
  const domains = opts.domains ?? [];
  if (domains.length > 0) {
    query.set("domains", domains.join(","));
  }
  query.set("limit", String(opts.limit ?? EVOLVE_LIMIT));
  // Sent only when it is asked for, so an ordinary sweep goes out as the
  // ordinary sweep it always was.
  if (opts.includeAcknowledged === true) {
    query.set("include_acknowledged", "true");
  }
  return readEvolveQueue(await api<unknown>(`/evolve?${query.toString()}`));
}

/** Where both acknowledgment verbs live, for one domain. */
function ackPath(domain: string): string {
  return `/domains/${encodeSegment(domain)}/evolve/ack`;
}

/**
 * The body both verbs take.
 *
 * A note that is only whitespace is left out rather than sent: it would be
 * stored in the engram's frontmatter, where a blank string is a line of noise
 * that reads as a reason somebody gave and did not.
 */
function ackBody(permalink: string, rule: string, note?: string): AckBody {
  const said = note?.trim() ?? "";
  return said === "" ? { permalink, rule } : { permalink, rule, note: said };
}

/**
 * Rule one finding intentional, so future sweeps stop raising it.
 *
 * The scope it holds for is never sent: the server runs detection for the
 * engram and takes the firing finding's own evidence, so neither a person nor
 * an agent ever handles a fingerprint. Acknowledging the same rule again
 * replaces the entry, which is what makes this the re-acknowledge call too.
 */
export async function acknowledgeFinding(
  domain: string,
  permalink: string,
  rule: string,
  note?: string,
): Promise<void> {
  await api(ackPath(domain), {
    method: "POST",
    body: JSON.stringify(ackBody(permalink, rule, note)),
  });
}

/** Take an acknowledgment back, leaving the engram's others alone. */
export async function unacknowledgeFinding(
  domain: string,
  permalink: string,
  rule: string,
): Promise<void> {
  await api(ackPath(domain), {
    method: "DELETE",
    body: JSON.stringify(ackBody(permalink, rule)),
  });
}

File: crates/service/src/rest/evolve.rs
//! The consolidation queue: what the knowledge needs next, ranked, in the shape
//! a maintenance page renders.
//!
//! The same sweep the `evolve` MCP tool runs, answered read-only. The tool
//! records that a sweep happened, because an agent that asks for the queue is
//! about to work it; this route calls the detection half instead
//! ([`crate::engine::Engine::evolve_detect`]), so a person leaving the page open
//! never tells this machine that its backlog was attended to. Nothing here
//! writes: not an engram, not the maintenance state a Stop hook nudges from.
//!
//! The engine's own JSON is handed back unchanged, like every other read on this
//! surface, so the page and the tool describe one queue rather than two.

use axum::Json;
use axum::extract::State;
use serde::Deserialize;
use serde_json::Value;
use utoipa::IntoParams;

use super::{ApiError, ApiQuery, ProblemDetail, RestState, csv};
use crate::params::EvolveParams;

/// The query string `GET /evolve` takes, mirroring [`EvolveParams`] minus its
/// `today`.
///
/// That omission is the one deliberate difference from the tool: pinning the
/// date makes an agent's run reproducible across a session, and a page always
/// asks about now. A caller who needs a fixed date has the tool.
#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub struct EvolveQuery {
    /// Restrict the sweep to these domains, comma separated. Defaults to every
    /// registered domain.
    #[serde(default)]
    #[param(example = "eng,ops")]
    domains: Option<String>,
    /// Restrict to these detector families, comma separated: `temporal`,
    /// `structure` or `redundancy`. Defaults to all three.
    #[serde(default)]
    #[param(example = "temporal,structure")]
    families: Option<String>,
    /// Restrict to these rule ids, comma separated, for example `V006,V201`.
    #[serde(default)]
    #[param(example = "V006")]
    rules: Option<String>,
    /// Drop findings scoring under this priority, 0 to 100.
    #[serde(default)]
    #[param(example = 50)]
    min_priority: Option<u8>,
    /// Page size. Defaults to 10, clamped by the engine to 100.
    #[serde(default)]
    #[param(example = 25)]
    limit: Option<usize>,
    /// One-based page number. Defaults to 1.
    #[serde(default)]
    #[param(example = 1)]
    page: Option<usize>,
    /// Include the findings acknowledgments suppressed, each marked
    /// `acknowledged` with the note that silenced it. Off by default: the queue
    /// is what still needs deciding.
    #[serde(default)]
    #[param(example = true)]
    include_acknowledged: Option<bool>,
}

/// `GET /evolve` - the consolidation queue: the ranked list of what the
/// registered domains need next, with the evidence behind each finding and the
/// instruction that says how to work it.
///
/// A read and only a read. Detection runs over the scope and nothing at all is
/// written, the maintenance state included: viewing the queue never counts as a
/// consolidation run, so opening this page does not stop a Stop hook from asking
/// for the sweep that is actually owed. The `evolve` MCP tool is the surface that
/// records a run, because an agent calling it is about to do the work.
///
/// Each row carries its rank across the whole result (`n`), its priority, the
/// rule that fired, the engram it fired on, and a `class`: `mechanical` for work
/// that completes intent the knowledge already records, `judgment` for work that
/// changes what it claims. A client renders the two differently - a judgment
/// finding is a question for a person, never a change to apply.
///
/// The per-rule instruction rides `actions` rather than a column, so a page of
/// findings from one rule carries it once; only the rules on this page appear.
/// `families` counts the whole filtered result rather than the page, which is
/// what section headings are drawn from, and `truncations` names any per-domain
/// cap that fired so a short queue is never mistaken for a finished one.
///
/// `acknowledged` counts what acknowledgments kept out of the queue, whole and
/// per family, so a short queue is never mistaken for a healthy one; a row whose
/// acknowledgment was given for evidence that has since changed comes back
/// carrying `ack_stale` and the old `ack_note`. Pass `include_acknowledged` to
/// see the suppressed rows themselves, each marked `acknowledged`.
///
/// `today` is not exposed. The temporal rules are evaluated as of now, which is
/// the only question a page asks; the tool takes a pinned date for a run that
/// has to be reproducible.
#[utoipa::path(
    get,
    path = "/api/v1/evolve",
    tag = "maintenance",
    operation_id = "get_evolve_queue",
    params(EvolveQuery),
    responses(
        (
            status = 200,
            description = "The engine's own evolve payload, unchanged: the swept \
                           scope, the ranked queue page, the per-family counts \
                           over the whole result, the per-rule instructions for \
                           the rules on this page, the shared guidance and any \
                           truncation that fired.",
            body = Object,
            example = json!({
                "scope": {
                    "domains": ["eng"],
                    "families": [],
                    "rules": [],
                    "min_priority": null,
                    "today": "2026-08-17"
                },
                "engrams_scanned": 42,
                "unparsed": 0,
                "total": 2,
                "page": 1,
                "limit": 10,
                "count": 2,
                "families": [{ "family": "temporal", "findings": 2 }],
                "acknowledged": {
                    "total": 1,
                    "by_family": { "temporal": 0, "structure": 1, "redundancy": 0 }
                },
                "queue": [{
                    "n": 1,
                    "priority": 58,
                    "rule": "V006",
                    "class": "judgment",
                    "domain": "eng",
                    "permalink": "human-capture",
                    "title": "Incident capture",
                    "line": null,
                    "finding": "captured by a person and never reviewed",
                    "evidence": "generated.by human:jordi; recorded 2026-08-16; no verified entry",
                    "fix": "review, then record a verified entry (edit_engram set_frontmatter verified)"
                }],
                "actions": [{
                    "rule": "V006",
                    "instruction": "A person captured this directly and nobody has reviewed it since. ..."
                }],
                "guidance": "This queue changes nothing by itself. ...",
                "truncations": []
            }),
        ),
        (
            status = 400,
            description = "The query string will not parse.",
            body = ProblemDetail,
            content_type = "application/problem+json",
        ),
        (
            status = 401,
            description = "No identity.",
            body = ProblemDetail,
            content_type = "application/problem+json",
        ),
        (
            status = 403,
            description = "The trusted-header identity names a disabled account.",
            body = ProblemDetail,
            content_type = "application/problem+json",
        ),
        (
            status = 404,
            description = "`domains` names a domain nobody registered.",
            body = ProblemDetail,
            content_type = "application/problem+json",
        ),
        (
            status = 422,
            description = "`families` or `rules` names something the catalog does \
                           not have.",
            body = ProblemDetail,
            content_type = "application/problem+json",
        ),
    ),
)]
pub async fn queue(
    State(state): State<RestState>,
    ApiQuery(query): ApiQuery<EvolveQuery>,
) -> Result<Json<Value>, ApiError> {
    let value = state
        .engine
        .evolve_detect(&EvolveParams {
            domains: csv(query.domains.as_deref()),
            families: csv(query.families.as_deref()),
            rules: csv(query.rules.as_deref()),
            min_priority: query.min_priority,
            limit: query.limit,
            page: query.page,
            include_acknowledged: query.include_acknowledged.unwrap_or(false),
            // Never from the caller: see the type and the operation doc above.
            today: None,
        })
        .await?;
    Ok(Json(value))
}
